backend/tournesol/views/unconnected_entities.py

```diff
@@ -28,54 +58,40 @@ class UnconnectedEntitiesView(PollScopedViewMixin, generics.ListAPIView):
-            all_connections[entity_1_id].add(entity_2_id)
-            all_connections[entity_2_id].add(entity_1_id)
+        neighbors = defaultdict(set)
+        for (entity_1_id, entity_2_id) in Comparison.objects.filter(
+            poll=self.poll_from_url, user=self.request.user
+        ).values_list("entity_1_id", "entity_2_id"):
+            neighbors[entity_1_id].add(entity_2_id)
+            neighbors[entity_2_id].add(entity_1_id)
 
-        def get_related_entities(entity_id):
-            already_visited_nodes = set()
-            related_entities = {entity_id}
-            to_visit = {entity_id}
+        distances = {source_node.id: 0}
+        bfs_queue = deque([(source_node.id, 0)])
 
-            while to_visit:
-                node = to_visit.pop()
-                already_visited_nodes.add(node)
-                connections = all_connections.get(node, set())
-                to_visit.update(connections - already_visited_nodes)
-                related_entities.update(connections)
-
-            return related_entities
-
-        user_related_entities = get_related_entities(source_node.id)
-        user_all_entities = set()
-
-        for (entity_1_id, entity_2_id) in comparisons:
-            user_all_entities.add(entity_1_id)
-            user_all_entities.add(entity_2_id)
-
-        # Order the entities by number of comparisons so that the logged user
-        # is invited to compare entities with the least comparisons first.
-        comparison_counts = (
-            Comparison.objects.filter(user=self.request.user)
-            .filter(Q(entity_1=OuterRef("pk")) | Q(entity_2=OuterRef("pk")))
-            .annotate(count=Func("id", function="Count"))
-            .values("count")
-        )
+        while bfs_queue:
+            node, distance = bfs_queue.popleft()
+            for next_node in neighbors[node]:
+                if next_node not in distances:
+                    distances[next_node] = distance+1
+                    bfs_queue.append((next_node, distance+1))
 
-        return (
-            Entity.objects.filter(id__in=user_all_entities - user_related_entities)
-            .annotate(n_comparisons=Subquery(comparison_counts))
-            .order_by("n_comparisons")
+        # The entities filtered are the entities which don't have a distance
+        # because they are unreachable if the strict option has been left to
+        # True (the default). Otherwise all the entities except the source are
+        # selected.
+        strict = self.request.query_params.get("strict") != "false"
+        entity_ids = set(e_id for e_id in neighbors if (
+            (e_id not in distances) if strict else (distances.get(e_id, 2) > 1)
+        ))
+        sorted_entity_ids = sorted(
+            entity_ids,
+            # Sorting first by distance and secondly by number of comparisons
+            key=lambda x: (-distances.get(x, math.inf), len(neighbors.get(x, [])))
```

Comment:
To ensure the sorted list is deterministic when requesting multiple pages:
```suggestion
            key=lambda x: (-distances.get(x, math.inf), len(neighbors.get(x, []), x))
```

Reply:
Fixed. Thanks!